src/rustllvm/PassWrapper.cpp

```diff
@@ -1496,6 +1498,39 @@ LLVMRustParseBitcodeForLTO(LLVMContextRef Context,
   return wrap(std::move(*SrcOrError).release());
 }
 
+// Find the bitcode section in the object file data and return it as a slice.
+// Fail if the bitcode section is present but empty.
+//
+// On success, the return value is the pointer to the start of the slice and
+// `out_len` is filled with the (non-zero) length. On failure, the return value
+// is `nullptr` and `out_len` is set to zero.
+extern "C" const char*
+LLVMRustGetBitcodeSliceFromObjectData(const char *data,
+                                      size_t len,
+                                      size_t *out_len) {
+  *out_len = 0;
+
+  StringRef Data(data, len);
+  MemoryBufferRef Buffer(Data, ""); // The id is unused.
+
+  Expected<std::unique_ptr<object::ObjectFile>> ObjectFileOrError =
+    object::ObjectFile::createObjectFile(Buffer);
+  if (!ObjectFileOrError) {
+    LLVMRustSetLastError(toString(ObjectFileOrError.takeError()).c_str());
+    return nullptr;
+  }
+
+  Expected<MemoryBufferRef> BitcodeOrError =
+    object::IRObjectFile::findBitcodeInObject(**ObjectFileOrError);
```

Comment:
FWIW I was poking at this recently and this can actually be simplified so simply call `findBitcodeInMemBuffer` which eschews object file parsing and also gets it working when the input slice of memory is actually already an IR object file. I believe that would make `-Clto` immediately compatible with `-C linker-plugin-lto`?

Reply:
Good suggestion, I have updated the relevant commit.